libcxx/include/__flat_map/flat_map.h

```diff
@@ -846,7 +846,7 @@ class flat_map {
         __compare_(std::forward<_CompArg>(__comp)...) {}
 
   _LIBCPP_HIDE_FROM_ABI bool __is_sorted_and_unique(auto&& __key_container) const {
-    auto __greater_or_equal_to = [this](const auto& __x, const auto& __y) { return !__compare_(__x, __y); };
+    auto __greater_or_equal_to = [this](const auto& __x, const auto& __y) -> bool { return !__compare_(__x, __y); };
```

Comment:
I _think_ it's expected that `decltype(__compare_(__x, __y))` already model _`boolean-testable`_ (or even is exactly `bool` in most normal usages).

If the standard were already so requiring, we wouldn't need to change these lambdas. But I haven't found such a requirement...

Reply:
Oh, I was likely to be wrong. But are such conversions already performed in algorithms?

Reply:
> Oh, I was likely to be wrong. But are such conversions already performed in algorithms?

https://godbolt.org/z/nbG9v4zqY

Reply:
Oh, I see. It's the nature of lambda expression without trailing return type that unfortunately decay-copies the result.

Perhaps there should be some test cases for this, with types in `boolean_testable.h` reused?

Reply:
> Perhaps there should be some test cases for this, with types in `boolean_testable.h` reused?

I don't have a local repository for llvm, co-working is welcome.

Reply:
I recalled that the current standard wording doesn't actually require the result type of `Predicate` of `erase_if` for `flat_meow` to model _`boolean-testable`_. The boolean-testablity requirements seem only imposed for `erase_if` for sequence containers.

For `flat_meow`, the current wording uses explicit cast to `bool`. So should we use `bool(__compare_(...))` instead, or open an LWG issue for this?

Reply:
> I recalled that the current standard wording doesn't actually require the result type of `Predicate` of `erase_if` for `flat_meow` to model _`boolean-testable`_. The boolean-testablity requirements seem only imposed for `erase_if` for sequence containers.
> 

I mainly referred to https://eel.is/c++draft/algorithms#requirements-6